livre/lib/data/RawDataSource.cpp

```diff
@@ -76,16 +101,36 @@ struct RawDataSource::Impl
         volInfo.overlap = Vector3ui( 0u );
         volInfo.rootNode = RootNode( 1, Vector3ui( 1 ));
         volInfo.maximumBlockSize = volInfo.voxels;
+
+        _inputType = volInfo.dataType;
+        const auto output = uri.findQuery( "output" );
+        if( output == uri.queryEnd( ))
+            _outputType = _inputType;
+        else
+        {
+            _outputType = getDataType( output->second );
+            volInfo.dataType = _outputType;
+        }
     }
 
     ~Impl() {}
 
     MemoryUnitPtr getData( const LODNode& node )
     {
-        const size_t dataSize = node.getBlockSize().product();
-        return MemoryUnitPtr(
-            new ConstMemoryUnit( _mmap.getAddress< uint8_t >() + _headerSize,
-                                 dataSize ));
+        const size_t size = node.getBlockSize().product();
+        const uint8_t* ptr = _mmap.getAddress< uint8_t >() + _headerSize;
+        if( _inputType == _outputType )
+            return  MemoryUnitPtr( new ConstMemoryUnit( ptr, size ));
+
+        // only unsigned integer conversions are supported!
+        if( _inputType == DT_UINT16 && _outputType == DT_UINT8 )
+            return _scale< uint8_t, uint16_t >( ptr, size );
+        if( _inputType == DT_UINT32 && _outputType == DT_UINT8 )
+            return _scale< uint8_t, uint32_t >( ptr, size );
+        if( _inputType == DT_UINT32 && _outputType == DT_UINT16 )
+            return _scale< uint16_t, uint32_t >( ptr, size );
```

Comment:
reverse the order: template< class O, class I > -> template< class I, class O> so that it reads in the logical order (uint32_t ->uint16_t ) ?